R/GroupSequentialDesign.R

```diff
@@ -0,0 +1,183 @@
+        c2_pivots = "numeric",
+        x1_norm_pivots = "numeric",
+        weights   = "numeric"
+    ),  contains = "TwoStageDesign"
+    )
+
+
+
+#' @param n1 cf. slot
+#' @param c1f cf. slot
+#' @param c1e cf. slot
+#' @param n2 cf. slot
+#' @param c2_pivots cf. slot
+#' @param x1_norm_pivots cf. slot
+#' @param weights cf. slot
+#' @param ... further optional arguments
+#'
+#' @rdname GSDesign-class
+#' @export
+GSDesign <- function(n1, c1f, c1e, n2, c2_pivots, x1_norm_pivots, weights) {
+    if (any(diff(sapply(list(c2_pivots, x1_norm_pivots, weights), length)) != 0))
+        stop("pivots and weights must all be of the same length")
+    if (any(x1_norm_pivots < -1) | any(x1_norm_pivots > 1))
+        stop("x1_norm_pivots must be in [-1, 1], is scaled automatically")
+    if (any(weights <= 0))
+        stop("weights must be positive")
+    new("GSDesign", n1 = n1, c1f = c1f, c1e = c1e, n2 = n2,
+        c2_pivots = c2_pivots, x1_norm_pivots = x1_norm_pivots, weights = weights)
+}
+
+
+
+#' @details GQDesign simply creates a TwoStageDesign object with a Gaussian
+#'     quadrature numerical integration rule.
+#'
+#' @param order order (i.e. number of pivot points in the interior of [c1f, c1e])
+#'     of the Gaussian quadrature rule to use for integration
+#'
+#' @rdname GSDesign-class
+#' @export
+GS_GQDesign <- function(n1, c1f, c1e, n2, c2_pivots, order) {
+    if (length(c2_pivots) != order)
+        stop("length of pivot vectors does not fit")
+    rule <- GaussLegendreRule(order)
+    GSDesign(n1 = n1, c1f = c1f, c1e = c1e, n2 = n2, c2_pivots = c2_pivots,
+             x1_norm_pivots = rule$nodes, weights = rule$weights)
+}
+
+
+
+#' @param x object to get parameters from
+#'
+#' @rdname GSDesign-class
+#' @export
+setMethod("tunable_parameters", signature("GSDesign"),
+          function(x, ...) c(x@n1, x@c1f, x@c1e, x@n2, x@c2_pivots))
+
+
+
+
+#' @param params vector of design parameters (must be in same order as returned
+#'     by \code{as.numeric(design)})
+#' @param object object to update
+#'
+#' @rdname GSDesign-class
+#' @export
+setMethod("update", signature("GSDesign"),
+    function(object, params, ...) {
+        k <- length(object@weights)
+        if( (length(params) - 4) != k)
+            stop("parameter length does not fit")
+        new("GSDesign",
+            n1  = params[1],
+            c1f = params[2],
+            c1e = params[3],
+            n2  = params[4],
+            c2_pivots = params[5:(length(params))],
+            x1_norm_pivots = object@x1_norm_pivots,
+            weights = object@weights)
+    })
+
+
+
+#' @param x1 stage-one outcome
+#' @param d design object
+#'
+#' @rdname GSDesign-class
+#' @export
+setMethod("n2", signature("GSDesign", "numeric"),
+          function(d, x1, ...) ifelse(x1 < d@c1f | x1 > d@c1e, 0, d@n2) )
+
+
+
+
+#' Convert a group-sequential design to a two-stage design
+#'
+#' @param d object of class \code{GSDesign}
+#'
+#' @export
+
+gs2ts <- function(d){
```

Comment:
I personally do not like cryptic function names - nobody understands what that is supposed to mean x)
The elegant way of solving the conversion problem is to make the constructor TwoStageDesign() itself a generic function. That way you can have it dispatch on different arguments (I did not need to do that so far because we essentially only ever had one constructor ;))

I.e. TwoStageDesign(gsdesign) would work and TwoStageDesign(n1, c1e, etc.) would also work ;)

Might be overkill and a workable solution with optional arguments might be easier, but this would be the most consitent way of solving this.